Unable to run graph/nweight in hibench 6.0

Granted this may be an issue with my cluster, or my configs, but since I've been working through a few issues trying to get 6.0 to run all tests on my system I thought it would be good to report this now while I continue to dive in.

```
start ScalaSparkNWeight bench
hdfs rm -r: /usr/hdp/current/hadoop-client/bin/hadoop --config /usr/hdp/current/hadoop-client/etc/hadoop fs -rm -r -skipTrash hdfs://192.168.0.1:8020/user/builder/HiBench/NWeight/Output
Deleted hdfs://192.168.0.1:8020/user/builder/HiBench/NWeight/Output
hdfs du -s: /usr/hdp/current/hadoop-client/bin/hadoop --config /usr/hdp/current/hadoop-client/etc/hadoop fs -du -s hdfs://192.168.0.1:8020/user/builder/HiBench/NWeight/Input
Export env: SPARKBENCH_PROPERTIES_FILES=/home/users/builder/cactus-test-framework/cactus-test-framework/workspace/plato-cactus-hibench_AQA-1256/hibench/report/nweight/spark/conf/sparkbench/sparkbench.conf
Submit Spark job: /opt/cray/spark2/default//bin/spark-submit  --properties-file /home/users/builder/cactus-test-framework/cactus-test-framework/workspace/plato-cactus-hibench_AQA-1256/hibench/report/nweight/spark/conf/sparkbench/spark.conf --class com.intel.hibench.sparkbench.graph.nweight.NWeight --master mesos://zk://zoo1:2181,zoo2:2181,zoo3:2181/mesos  /home/users/builder/cactus-test-framework/cactus-test-framework/workspace/plato-cactus-hibench_AQA-1256/hibench/sparkbench/assembly/target/sparkbench-assembly-6.0-SNAPSHOT-dist.jar hdfs://192.168.0.1:8020/user/builder/HiBench/NWeight/Input hdfs://192.168.0.1:8020/user/builder/HiBench/NWeight/Output 3 30 12 7 false graphx
16/11/10 12:57:38 WARN util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
16/11/10 12:57:40 WARN shortcircuit.DomainSocketFactory: The short-circuit local reads feature cannot be used because libhadoop cannot be loaded.
16/11/10 12:57:46 WARN scheduler.TaskSetManager: Lost task 9.0 in stage 5.0 (TID 67, nid00029.local): java.lang.NullPointerException
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 2.0 in stage 5.0 (TID 61) on executor nid00040.local: java.lang.NullPointerException (null) [duplicate 1]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 10.0 in stage 5.0 (TID 64) on executor nid00033.local: java.lang.NullPointerException (null) [duplicate 2]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 0.0 in stage 5.0 (TID 65) on executor nid00043.local: java.lang.NullPointerException (null) [duplicate 3]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 1.0 in stage 5.0 (TID 60) on executor nid00013.local: java.lang.NullPointerException (null) [duplicate 4]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 6.0 in stage 5.0 (TID 68) on executor nid00037.local: java.lang.NullPointerException (null) [duplicate 5]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 11.0 in stage 5.0 (TID 71) on executor nid00037.local: java.lang.NullPointerException (null) [duplicate 6]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 3.0 in stage 5.0 (TID 66) on executor nid00019.local: java.lang.NullPointerException (null) [duplicate 7]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 5.0 in stage 5.0 (TID 62) on executor nid00039.local: java.lang.NullPointerException (null) [duplicate 8]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 8.0 in stage 5.0 (TID 69) on executor nid00044.local: java.lang.NullPointerException (null) [duplicate 9]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 9.1 in stage 5.0 (TID 73) on executor nid00029.local: java.lang.NullPointerException (null) [duplicate 10]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 7.0 in stage 5.0 (TID 70) on executor nid00022.local: java.lang.NullPointerException (null) [duplicate 11]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 10.1 in stage 5.0 (TID 74) on executor nid00033.local: java.lang.NullPointerException (null) [duplicate 12]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 2.1 in stage 5.0 (TID 72) on executor nid00040.local: java.lang.NullPointerException (null) [duplicate 13]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 0.1 in stage 5.0 (TID 75) on executor nid00043.local: java.lang.NullPointerException (null) [duplicate 14]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 4.0 in stage 5.0 (TID 63) on executor nid00017.local: java.lang.NullPointerException (null) [duplicate 15]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 1.1 in stage 5.0 (TID 76) on executor nid00013.local: java.lang.NullPointerException (null) [duplicate 16]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 3.1 in stage 5.0 (TID 79) on executor nid00019.local: java.lang.NullPointerException (null) [duplicate 17]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 6.1 in stage 5.0 (TID 78) on executor nid00037.local: java.lang.NullPointerException (null) [duplicate 18]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 11.1 in stage 5.0 (TID 77) on executor nid00037.local: java.lang.NullPointerException (null) [duplicate 19]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 5.1 in stage 5.0 (TID 80) on executor nid00039.local: java.lang.NullPointerException (null) [duplicate 20]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 9.2 in stage 5.0 (TID 82) on executor nid00029.local: java.lang.NullPointerException (null) [duplicate 21]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 8.1 in stage 5.0 (TID 81) on executor nid00044.local: java.lang.NullPointerException (null) [duplicate 22]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 10.2 in stage 5.0 (TID 84) on executor nid00033.local: java.lang.NullPointerException (null) [duplicate 23]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 7.1 in stage 5.0 (TID 83) on executor nid00022.local: java.lang.NullPointerException (null) [duplicate 24]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 4.1 in stage 5.0 (TID 87) on executor nid00017.local: java.lang.NullPointerException (null) [duplicate 25]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 2.2 in stage 5.0 (TID 85) on executor nid00040.local: java.lang.NullPointerException (null) [duplicate 26]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 0.2 in stage 5.0 (TID 86) on executor nid00043.local: java.lang.NullPointerException (null) [duplicate 27]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 1.2 in stage 5.0 (TID 88) on executor nid00013.local: java.lang.NullPointerException (null) [duplicate 28]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 6.2 in stage 5.0 (TID 91) on executor nid00037.local: java.lang.NullPointerException (null) [duplicate 29]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 11.2 in stage 5.0 (TID 90) on executor nid00037.local: java.lang.NullPointerException (null) [duplicate 30]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 3.2 in stage 5.0 (TID 89) on executor nid00019.local: java.lang.NullPointerException (null) [duplicate 31]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 9.3 in stage 5.0 (TID 93) on executor nid00029.local: java.lang.NullPointerException (null) [duplicate 32]
16/11/10 12:57:46 ERROR scheduler.TaskSetManager: Task 9 in stage 5.0 failed 4 times; aborting job
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 5.2 in stage 5.0 (TID 92) on executor nid00039.local: java.lang.NullPointerException (null) [duplicate 33]
16/11/10 12:57:46 INFO scheduler.TaskSetManager: Lost task 8.2 in stage 5.0 (TID 94) on executor nid00044.local: java.lang.NullPointerException (null) [duplicate 34]
16/11/10 12:57:46 INFO scheduler.StatsReportListener: Finished stage: Stage(5, 0); Name: 'mapPartitions at GraphImpl.scala:207'; Status: failed(Job aborted due to stage failure: Task 9 in stage 5.0 failed 4 times, most recent failure: Lost task 9.3 in stage 5.0 (TID 93, nid00029.local): java.lang.NullPointerException
Exception in thread "main" 16/11/10 12:57:46 INFO scheduler.StatsReportListener: remote bytes read:(count: 33, mean: 1763.363636, stdev: 62.868221, max: 1793.000000, min: 1630.000000)
org.apache.spark.SparkException: Job aborted due to stage failure: Task 9 in stage 5.0 failed 4 times, most recent failure: Lost task 9.3 in stage 5.0 (TID 93, nid00029.local): java.lang.NullPointerException
Caused by: java.lang.NullPointerException
16/11/10 12:57:46 WARN scheduler.TaskSetManager: Lost task 10.3 in stage 5.0 (TID 95, nid00033.local): TaskKilled (killed intentionally)
16/11/10 12:57:46 WARN scheduler.TaskSetManager: Lost task 7.2 in stage 5.0 (TID 96, nid00022.local): TaskKilled (killed intentionally)
16/11/10 12:57:46 WARN scheduler.TaskSetManager: Lost task 0.3 in stage 5.0 (TID 99, nid00043.local): TaskKilled (killed intentionally)
16/11/10 12:57:46 ERROR server.TransportRequestHandler: Error while invoking RpcHandler#receive() for one-way message.
org.apache.spark.SparkException: Could not find CoarseGrainedScheduler.
16/11/10 12:57:46 ERROR server.TransportRequestHandler: Error while invoking RpcHandler#receive() for one-way message.
org.apache.spark.SparkException: Could not find CoarseGrainedScheduler.
16/11/10 12:57:46 ERROR server.TransportRequestHandler: Error while invoking RpcHandler#receive() for one-way message.
org.apache.spark.SparkException: Could not find CoarseGrainedScheduler.
16/11/10 12:57:46 ERROR server.TransportRequestHandler: Error while invoking RpcHandler#receive() for one-way message.
org.apache.spark.SparkException: Could not find CoarseGrainedScheduler.
16/11/10 12:57:47 INFO util.ShutdownHookManager: Deleting directory /var/spark/tmp/spark-434eb61a-ac60-48ea-a3f8-01dce3f19c17
ERROR: Spark job com.intel.hibench.sparkbench.graph.nweight.NWeight failed to run successfully.
Hint: You can goto /home/users/builder/cactus-test-framework/cactus-test-framework/workspace/plato-cactus-hibench_AQA-1256/hibench/report/nweight/spark/conf/../bench.log to check for detailed log.
Opening log tail for you:

I1110 12:57:47.526414 47918 sched.cpp:1143] Stopping framework 'afe4dd32-b0f4-42a4-a07d-0ad05b9c8328-0208'
16/11/10 12:57:47 INFO mesos.MesosCoarseGrainedSchedulerBackend: driver.run() returned with code DRIVER_STOPPED
16/11/10 12:57:47 INFO spark.MapOutputTrackerMasterEndpoint: MapOutputTrackerMasterEndpoint stopped!
16/11/10 12:57:47 INFO memory.MemoryStore: MemoryStore cleared
16/11/10 12:57:47 INFO storage.BlockManager: BlockManager stopped
16/11/10 12:57:47 INFO storage.BlockManagerMaster: BlockManagerMaster stopped
16/11/10 12:57:47 INFO scheduler.OutputCommitCoordinator$OutputCommitCoordinatorEndpoint: OutputCommitCoordinator stopped!
16/11/10 12:57:47 INFO spark.SparkContext: Successfully stopped SparkContext
16/11/10 12:57:47 INFO util.ShutdownHookManager: Shutdown hook called
16/11/10 12:57:47 INFO util.ShutdownHookManager: Deleting directory /var/spark/tmp/spark-434eb61a-ac60-48ea-a3f8-01dce3f19c17
ERROR: graph/nweight/spark failed to run successfully.
```

I think we need more information, @carsonwang why do you suspect it involves mapReduceTriplets and aggregateMessages?

I see the [change ]( https://github.com/intel-hadoop/HiBench/pull/262/files)I made has already been merged into [master](https://github.com/intel-hadoop/HiBench/blob/master/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/GraphxNWeight.scala), mapReduceTriplets is replaced with aggregateMessages already and we don't see a compile error reported

@harschware do you have your HiBench properties file (removing anything sensitive of course) and any other relevant stdout to share with us, let's say from your work directory (master/worker/application output) and/or your logs directory?

I do see the following which could be a clue, what exactly is null that we're trying to access?

```
16/11/10 12:57:46 INFO scheduler.StatsReportListener: Finished stage: Stage(5, 0); Name: '**mapPartitions at GraphImpl.scala:207'**; Status: failed(Job aborted due to stage failure: Task 9 in stage 5.0 failed 4 times, most recent failure: Lost task 9.3 in stage 5.0 (TID 93, nid00029.local): java.lang.NullPointerException
Exception in thread "main" 16/11/10 12:57:46 INFO scheduler.StatsReportListener: remote bytes read:(count: 33, mean: 1763.363636, stdev: 62.868221, max: 1793.000000, min: 1630.000000)
org.apache.spark.SparkException: Job aborted due to stage failure: Task 9 in stage 5.0 failed 4 times, most recent failure: Lost task 9.3 in stage 5.0 (TID 93, nid00029.local): **java.lang.NullPointerException
Caused by: java.lang.NullPointerException**
```

 then lots of null mentioning lines before it and after this:
```
16/11/10 12:57:40 WARN shortcircuit.DomainSocketFactory: The short-circuit local reads feature cannot be used because libhadoop cannot be loaded.
16/11/10 12:57:46 WARN scheduler.TaskSetManager: Lost task 9.0 in stage 5.0 (TID 67, nid00029.local): java.lang.NullPointerException
```

--- a/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/Driver.scala
+++ b/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/Driver.scala
@@ -19,7 +19,6 @@ package com.intel.hibench.sparkbench.graph.nweight
 
 import org.apache.spark.{SparkContext, SparkConf}
 import org.apache.spark.storage.StorageLevel
-import org.apache.spark.scheduler.StatsReportListener
 
 /** 
  * Compute NWeight for Graph G(V, E) as defined below
@@ -75,12 +74,12 @@ object NWeight extends Serializable{
       sparkConf.setAppName("NWeightPregel")
     val sc = new SparkContext(sparkConf)
 
-    sc.addSparkListener(new StatsReportListener)
-
     if (model.toLowerCase == "graphx") {
       GraphxNWeight.nweight(sc, input, output, step, maxDegree, numPartitions, storageLevel)
     } else {
       PregelNWeight.nweight(sc, input, output, step, maxDegree, numPartitions, storageLevel)
     }
+
+    sc.stop()
   }
 }

--- a/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/Driver.scala
+++ b/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/Driver.scala
@@ -20,14 +20,18 @@ package com.intel.hibench.sparkbench.graph.nweight
 import org.apache.spark.{SparkContext, SparkConf}
 import org.apache.spark.storage.StorageLevel
 
-/** 
- * Compute NWeight for Graph G(V, E) as defined below
- *     Weight(1)(u, v) = edge(u, v)
- *     Weight(n)(u, v) = Sum (over {x|there are edges (u, x) and (x, v)}) Weight(n-1)(u, x)*Weight(1)(x, v)
- * 
- * Input is given in Text file format. Each line represents a Node and all out edges of that node (edge weight specified) 
- *  <vertex> <vertex1>:<weight1>, <vertex2>:<weight2> ...) 
+/**
+ * Compute NWeight for Graph G(V, E) as defined below.
+ *
+ * Weight(1)(u, v) = edge(u, v)
+ * Weight(n)(u, v) =
+ *   Sum (over {x|there are edges (u, x) and (x, v)}) Weight(n-1)(u, x) * Weight(1)(x, v)
+ *
+ * Input is given in Text file format. Each line represents a Node and all out edges of that node
+ * (edge weight specified)
+ * <vertex> <vertex1>:<weight1>,<vertex2>:<weight2> ...)
  */
+
 object NWeight extends Serializable{
  
   def parseArgs(args: Array[String]) = {

--- a/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/GraphxNWeight.scala
+++ b/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/GraphxNWeight.scala
@@ -86,7 +86,7 @@ object GraphxNWeight extends Serializable{
       (e.srcId, (e.dstId, e.attr))
     }.groupByKey(part).map { case (id, seq) =>
       val vdata = new SizedPriorityQueue(maxDegree)
-      seq.foreach(vdata.enqueue(_))
+      seq.foreach(vdata.enqueue)
       (id, vdata)
     }
 
@@ -94,8 +94,8 @@ object GraphxNWeight extends Serializable{
 
     var msg: RDD[(VertexId, Long2DoubleOpenHashMap)] = null
     for (i <- 2 to step) {
-      msg = g.aggregateMessages(mapF _, reduceF _)
-      g = g.outerJoinVertices(msg)(updateF _).persist(storageLevel)
+      msg = g.aggregateMessages(mapF, reduceF)
+      g = g.outerJoinVertices(msg)(updateF).persist(storageLevel)
     }
 
     g.vertices.map { case (vid, vdata) => 

--- a/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/GraphxNWeight.scala
+++ b/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/GraphxNWeight.scala
@@ -30,7 +30,7 @@ import it.unimi.dsi.fastutil.longs.Long2DoubleOpenHashMap
  *     Weight(n)(u, v) = Sum (over {x|there are edges (u, x) and (x, v)}) Weight(n-1)(u, x)*Weight(1)(x, v)
  *
  * Input is given in Text file format. Each line represents a Node and all out edges of that node (edge weight specified)
- *  <vertex> <vertex1>:<weight1>, <vertex2>:<weight2> ...)
+ *  <vertex> <vertex1>:<weight1>,<vertex2>:<weight2> ...)
  */
 
 object GraphxNWeight extends Serializable{
@@ -43,7 +43,7 @@ object GraphxNWeight extends Serializable{
       if (target != id)
         theMap.put(target, wn * edgeAttribute)
     }
-    Iterator((id, theMap))
+    edge.sendToSrc(theMap)
   }
 
   def reduceF(c1: Long2DoubleOpenHashMap, c2: Long2DoubleOpenHashMap) = {
@@ -94,7 +94,7 @@ object GraphxNWeight extends Serializable{
 
     var msg: RDD[(VertexId, Long2DoubleOpenHashMap)] = null
     for (i <- 2 to step) {
-      msg = g.aggregateMessages(mapF _, reduceF _, TripletFields.Src)
+      msg = g.aggregateMessages(mapF _, reduceF _)
       g = g.outerJoinVertices(msg)(updateF _).persist(storageLevel)
     }
 

--- a/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/NWeightDataGenerator.scala
+++ b/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/NWeightDataGenerator.scala
@@ -106,6 +106,8 @@ object NWeightDataGenerator {
     }
 
     resultData.saveAsTextFile(outputPath)
+
+    sc.stop()
   }
 
   def loadModel(modelPath: String, sc: SparkContext, partitions: Int): MatrixFactorizationModel = {

--- a/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/PregelNWeight.scala
+++ b/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/PregelNWeight.scala
@@ -34,7 +34,7 @@ import it.unimi.dsi.fastutil.longs.Long2DoubleOpenHashMap
  *
  * Input is given in Text file format. Each line represents a Node and all out edges of that node
  * (edge weight specified)
- *  <vertex> <vertex1>:<weight1>, <vertex2>:<weight2> ...)
+ *  <vertex> <vertex1>:<weight1>,<vertex2>:<weight2> ...)
  */
 
 object PregelNWeight extends Serializable{

--- a/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/PregelNWeight.scala
+++ b/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/PregelNWeight.scala
@@ -34,7 +34,7 @@ import it.unimi.dsi.fastutil.longs.Long2DoubleOpenHashMap
  *
  * Input is given in Text file format. Each line represents a Node and all out edges of that node
  * (edge weight specified)
- *  <vertex> <vertex1>:<weight1>,<vertex2>:<weight2> ...)
+ * <vertex> <vertex1>:<weight1>,<vertex2>:<weight2> ...)
  */
 
 object PregelNWeight extends Serializable{

--- a/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/PregelNWeight.scala
+++ b/sparkbench/graph/src/main/scala/com/intel/hibench/sparkbench/graph/nweight/PregelNWeight.scala
@@ -92,7 +92,7 @@ object PregelNWeight extends Serializable{
     var g = GraphImpl(edges, new SizedPriorityQueue(maxDegree), storageLevel, storageLevel).cache()
 
     g = Pregel(g, new Long2DoubleOpenHashMap, step, EdgeDirection.In)(
-      vProg _, sendMsg _, mergMsg _)
+      vProg, sendMsg, mergMsg)
 
     g.vertices.map { case (vid, vdata) => 
       var s = new StringBuilder

--- a/travis/benchmarks.lst
+++ b/travis/benchmarks.lst
@@ -2,3 +2,4 @@ micro.sort
 ml.bayes
 websearch.nutchindexing
 sql.scan
+graph.nweight

